proposals/1406_multiple-idps/README.md

```diff
@@ -0,0 +1,493 @@
+   the identity provider names and types are already stored in user sessions for use during refreshes. This code already
+   has sufficient protections to ensure that we can never accidentally use a different identity provider during refresh
+   compared to which was used during initial login, even if there is an accidental name collision (via UID comparisons).
+2. Pre-existing kubeconfigs would now refer to the wrong identity provider name, and would need to be regenerated.
+
+If an admin wants to add a pre-existing identityProvider to a pre-existing FederationDomain without interrupting
+pre-existing sessions or needing to generate new kubeconfigs, they could take care to make the "displayName" of
+the identity provider exactly match the name of the identity provider CR.
+
+#### Tests
+
+Lots of new unit and integration tests will be required for using multiple FederationDomains, multiple identity
+providers, and identity transformations and policies.
+
+#### New Dependencies
+
+https://github.com/google/cel-go would move from being an indirect dependency (via k8s libraries) to a direct dependency.
+
+#### Performance Considerations
+
+No problems are anticipated. CEL is up to the task from a performance point of view.
+
+#### Observability Considerations
+
+The status of FederationDomains will be updated to show new types of validation errors. Unexpected transformation errors
+during login attempts will be logged in the Pod logs.
+
+#### Security Considerations
+
+FederationDomains were already designed to securely control authentication into Kubernetes clusters. Allowing multiple
+sources of identity on a FederationDomain does not change that, except for allowing more potential users. See above for
+detailed discussion of identity conflict considerations on those additional users. Adding identity transformations and
+policies gives the admin more control over how the identities extracted from external identity providers are
+projected into Kubernetes.
+
+#### Usability Considerations
+
+This proposal does not change the user experience for the end user (kubectl user). This proposal does not include
+any changes to their kubeconfig or to the Pinniped CLI.
+
+This proposal adds more powerful configuration options for the Supervisor admin. By choosing CEL, we hope that the
+identity transforms and policies are simple for the admin to create, and are done in a language with which they might
+already be familiar due to its usage in Kubernetes. By allowing the admin to configure "examples" on the
+FederationDomain we hope to reduce the possibility of admins making programming mistakes in CEL expressions. Admins will
+need to understand how to anticipate and resolve identity conflicts, which is a new usability concern that we intend to
+address with documentation.
+
+#### Documentation Considerations
+
+See "Implementation Plan" section below.
+
+### Other Approaches Considered
```

Comment:
I wonder if it's worth mentioning the approach of having multiple IDPs with defaulted behavior. In a sense Pinniped would provide its own opinions on how to distinguish usernames and groups from different identity providers. Likely this would mean something like usernames are given a prefix of `<idp-id>:` and groups are passed through transparently.

Another approach could be to use the behavior above when the user does not specify their own transformations.

Reply:
If we feel that the implementation path for everything in this document is "too big" we could choose to do something like the above as an intermediate release until transformations are available.

Or we could implement transformations first and then multiple IDPs. Don't think we have the data to support this yet.

Reply:
We decided to implement the whole proposal.